File: tools/auditproof-verifier-go/canonical_recompute.go
// Canonical-hash recompute + Ed25519 signature verification — the Go mirror of
// the Rust verifiersrc/canonical_recompute.rs`.
//
// This is what makes stages 5-7 real. Reproducing the 8-step chain (stages 1-4)
// proves the chain is internally consistent; it says nothing about whether the
// document was signed. A forged proof carrying a chain it computed itself walks
// stages 1-4 perfectly. Only recomputing the signed message and checking the
// Ed25519 signature over it distinguishes a genuine AuditProof from a fabricated
// one.
//
// ## Which message is signed (mirrors the hosted verification endpoint)
//
//   - 1.0                    → final_hash (ASCII hex, prefix-stripped)
//   - 2.0                    → document_hash
//   - 2.1 + nanorix_only     → the recomputed the specification Part-3 canonical hash
//   - 2.1 + dual_signature   → not verifiable by this build → Absent
//   - 2.1 + tee_attested     → not verifiable by this build → Absent
//
// Recomputing the canonical hash from the document (rather than trusting the
// embedded `canonical_hash` field) is what catches canonical-view drift: a
// tampered `jurisdiction` / `org_id` / `capsule_id` never touches the 8-step
// chain, so it is invisible to stages 1-4 and shows up only here.
//
// ## Trust scope
//
// The signature is checked against the public key EMBEDDED in the proof. That
// proves INTEGRITY — the document has not been altered since it was signed — and
// is stage 7. It does NOT prove the key belongs to Nanorix; binding the key to a
// Nanorix-rooted trust anchor is the trust-chain manifest path (stage 8), which
// the Rust verifier implements via `VerifierPolicy::trust_chain` and this Go
// build does not yet carry.
//
// Forever-Standard discipline (the Forever-Standard wire discipline): the canonical view's field set,
// wire-name mapping, and the two server-side transforms (signing_key_version
// String → i64; the attestation subset) are part of the attestation contract.
// Any change here must keep the 100-fixture corpus at 100/100.

package auditproof

import (
	"bytes"
	"crypto/ed25519"
	"crypto/sha512"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"strconv"
	"strings"
)

// unsupportedModeSentinel marks a signedMessage result as "this build cannot
// verify the declared signing_mode" rather than a message to sign over. The NUL
// prefix can never occur in a hex digest. Mirrors UNSUPPORTED_MODE_SENTINEL in
// the Rust verifier.
const unsupportedModeSentinel = "\x00unsupported-signing-mode:"

// SignatureCheckKind is the outcome of the signature stage (stages 5-7).
type SignatureCheckKind int

const (
	// SignatureVerified — the signature checked out against the embedded key
	// over the correct per-version message.
	SignatureVerified SignatureCheckKind = iota

	// SignatureAbsent — nothing to check: no signature or key present. The
	// caller keeps the honest stage-4 "chain verified, signature NOT checked"
	// verdict rather than claiming a verification it did not perform.
	SignatureAbsent

	// SignatureFailed — a signature was present and did not verify.
	SignatureFailed

	// SignatureUnsupported — the document declares a signing_mode this build
	// cannot verify. Distinct from SignatureAbsent on purpose: signing_mode is
	// inside the canonical hash and is attacker-controllable, so if an
	// unrecognised mode produced the same verdict as a missing signature,
	// flipping the field would convert a rejection into reassurance — a
	// downgrade oracle. Mirrors SignatureCheck::Unsupported in the Rust verifier.
	SignatureUnsupported
)

// SignatureCheck pairs the outcome with the closed-set reason on failure.
type SignatureCheck struct {
	Kind   SignatureCheckKind
	Reason SignatureFailureReason
	// Mode is set only for SignatureUnsupported: the signing_mode this build
	// cannot verify, surfaced as algorithm_unsupported{found:"signing_mode=..."}.
	Mode string
}

// RecomputeCanonicalHash rebuilds the specification Part-3 canonical view from a
// proof's JSON and returns its RFC-8785 JCS SHA-512 hex digest — byte-identical
// to the server's `FullCdp::canonical_hash()` and to the Rust verifier's
// `recompute_canonical_hash`. Lowercase 128-char hex, or empty string on the
// impossible JCS-serialize failure so the signature check fails closed.
//
// `proof` must be a tree parsed with `json.Decoder.UseNumber()` so numeric
// literals keep their source form; see canonicalProofTree.
func RecomputeCanonicalHash(proof map[string]interface{}) string {
	view := map[string]interface{}{}

	// Always-present scalars (FullCdp wire name → canonical-view key). An
	// absent key canonicalizes to null, matching serde's behaviour for a field
	// with no skip attribute.
	view["version"] = proof["cdp_version"]
	view["signing_mode"] = proof["signing_mode"]
	view["jurisdiction"] = proof["jurisdiction"]
	view["authority_id"] = proof["authority_id"]

	// signing_key_version: FullCdp stores a String; the canonical view emits an
	// integer (server parses; unparseable → 0).
	skv := int64(0)
	if s, ok := proof["signing_key_version"].(string); ok {
		if parsed, err := strconv.ParseInt(s, 10, 64); err == nil {
			skv = parsed
		}
	}
	view["signing_key_version"] = json.Number(strconv.FormatInt(skv, 10))

	view["capsule_id"] = proof["capsule_id"]

	// org_id defaults to "" on the server (#[serde(default)] String).
	if v, ok := proof["org_id"]; ok {
		view["org_id"] = v
	} else {
		view["org_id"] = ""
	}

	// skip_serializing_if = Option::is_none → OMIT when absent or null.
	insertIfPresent(view, "parent_audit_proof_id", proof)
	insertIfPresent(view, "cdp_kind", proof)

	// Arrays carried verbatim (canonical-view key differs from wire name).
	if v, ok := proof["activity"]; ok && v != nil {
		view["activity_trail"] = v
	} else {
		view["activity_trail"] = []interface{}{}
	}
	if v, ok := proof["chain"]; ok && v != nil {
		view["destruction_chain"] = v
	} else {
		view["destruction_chain"] = []interface{}{}
	}

	view["destruction_state"] = proof["destruction_state"]

	// No skip attribute → serialized as null when absent (NOT omitted).
	view["destruction_failure_step"] = proof["destruction_failure_step"]

	insertIfPresent(view, "parent_proofs_merkle_root", proof)
	insertIfPresent(view, "record_receipts_merkle_root", proof)

	// No skip attribute → null when absent.
	view["runtime_attestation"] = proof["runtime_attestation"]

	// attestation subset. An empty fingerprint canonicalizes to null, matching
	// the server's `if fingerprint.is_empty() { None }`.
	att := map[string]interface{}{
		"timestamp_attestation": proof["timestamp_attestation"],
	}
	if s, ok := proof["attestation_chain_fingerprint"].(string); ok && s != "" {
		att["attestation_chain_fingerprint"] = s
	} else {
		att["attestation_chain_fingerprint"] = nil
	}
	view["attestation"] = att

	view["hash_algorithm"] = proof["hash_algorithm"]
	view["signature_algorithm"] = proof["signature_algorithm"]

	var buf bytes.Buffer
	if err := jcsEmit(&buf, view); err != nil {
		// A map of JSON-derived values always serializes; on the impossible
		// failure return empty so the comparison fails closed.
		return ""
	}
	sum := sha512.Sum512(buf.Bytes())
	return hex.EncodeToString(sum[:])
}

func insertIfPresent(view map[string]interface{}, key string, proof map[string]interface{}) {
	if v, ok := proof[key]; ok && v != nil {
		view[key] = v
	}
}

// canonicalProofTree re-parses the raw document with UseNumber so numeric
// literals keep their exact source form through JCS. The verification ladder
// itself works on a conventional `json.Unmarshal` tree (float64 numbers); only
// the canonical recompute needs the number-preserving view, because there the
// bytes are the cryptographic message.
func canonicalProofTree(jsonBytes []byte) (map[string]interface{}, bool) {
	dec := json.NewDecoder(bytes.NewReader(jsonBytes))
	dec.UseNumber()
	var v map[string]interface{}
	if err := dec.Decode(&v); err != nil {
		return nil, false
	}
	return v, true
}

// signedMessage selects the message a proof's signature covers, by version and
// signing mode. Returns ok=false when this build cannot determine the message,
// which the caller reports as Absent rather than as a failure.
func signedMessage(proof map[string]interface{}, cdpVersion string) (string, bool) {
	signingMode := "nanorix_only"
	if s, ok := proof["signing_mode"].(string); ok {
		signingMode = s
	}

	switch cdpVersion {
	case "1.0":
		s, _ := proof["final_hash"].(string)
		return StripHashPrefix(s), true
	case "2.0":
		s, _ := proof["document_hash"].(string)
		return StripHashPrefix(s), true
	case "2.1":
		if signingMode == "nanorix_only" {
			return RecomputeCanonicalHash(proof), true
		}
		// Any other declared mode is one this build cannot verify. NOT the same
		// as "no signature" — signalled with a sentinel the callers translate
		// into SignatureUnsupported, keeping this function's (string, bool) shape.
		return unsupportedModeSentinel + signingMode, true
	default:
		return "", false
	}
}

// verifyMessageWithKey decodes the base64 Ed25519 signature + public key and
// verifies `message` under them.
//
// Failure-reason ordering mirrors the Rust verifier: the signature is decoded
// and length-checked BEFORE the public key, so a proof with both malformed
// reports `malformed`, not `public_key_malformed`.
//
// One documented boundary: ed25519-dalek rejects a structurally invalid
// 32-byte public key at parse time (→ public_key_malformed), while Go's
// standard library has no exported point-validation and reports the same input
// as `does_not_verify`. Both reject; only the sub-reason can differ, and only
// for a key that is 32 bytes but not a valid curve point. Closing that gap
// would require a third-party curve library, which this module deliberately
// does not carry (see go.mod).
func verifyMessageWithKey(message, sigB64, pubB64 string) SignatureCheck {
	sig, err := base64.StdEncoding.DecodeString(StripBase64Prefix(sigB64))
	if err != nil || len(sig) != ed25519.SignatureSize {
		return SignatureCheck{Kind: SignatureFailed, Reason: SigMalformed}
	}
	pub, err := base64.StdEncoding.DecodeString(StripBase64Prefix(pubB64))
	if err != nil || len(pub) != ed25519.PublicKeySize {
		return SignatureCheck{Kind: SignatureFailed, Reason: SigPublicKeyMalformed}
	}
	if !ed25519.Verify(ed25519.PublicKey(pub), []byte(message), sig) {
		return SignatureCheck{Kind: SignatureFailed, Reason: SigDoesNotVerify}
	}
	return SignatureCheck{Kind: SignatureVerified}
}

// embeddedSignature returns the proof's attestation signature, if present and
// non-empty.
func embeddedSignature(proof map[string]interface{}) (string, bool) {
	s := lookupStringPath(proof, "attestation", "signature")
	if s == nil || *s == "" {
		return "", false
	}
	return *s, true
}

// VerifySignature verifies the proof's signature against the public key
// EMBEDDED in its attestation. Proves integrity (not tampered since signing),
// not authenticity. Returns Absent when no signature/key is present or the
// version/mode is not signature-verifiable here.
func VerifySignature(proof map[string]interface{}, cdpVersion string) SignatureCheck {
	pub := lookupStringPath(proof, "attestation", "public_key")
	if pub == nil || *pub == "" {
		pub = lookupStringPath(proof, "attestation", "verification_key")
	}
	if pub == nil || *pub == "" {
		return SignatureCheck{Kind: SignatureAbsent}
	}
	sig, ok := embeddedSignature(proof)
	if !ok {
		return SignatureCheck{Kind: SignatureAbsent}
	}
	message, ok := signedMessage(proof, cdpVersion)
	if !ok {
		return SignatureCheck{Kind: SignatureAbsent}
	}
	if mode, found := strings.CutPrefix(message, unsupportedModeSentinel); found {
		return SignatureCheck{Kind: SignatureUnsupported, Mode: mode}
	}
	return verifyMessageWithKey(message, sig, *pub)
}

// VerifySignatureAgainst verifies the proof's signature against a caller-
// supplied public key rather than the embedded one — the shape the trust-chain
// (stage 8) path uses to reject a forged proof that carries its own key.
func VerifySignatureAgainst(proof map[string]interface{}, cdpVersion, pubB64 string) SignatureCheck {
	sig, ok := embeddedSignature(proof)
	if !ok {
		return SignatureCheck{Kind: SignatureAbsent}
	}
	message, ok := signedMessage(proof, cdpVersion)
	if !ok {
		return SignatureCheck{Kind: SignatureAbsent}
	}
	if mode, found := strings.CutPrefix(message, unsupportedModeSentinel); found {
		return SignatureCheck{Kind: SignatureUnsupported, Mode: mode}
	}
	return verifyMessageWithKey(message, sig, pubB64)
}

// ─────────────────────────────────────────────────────────────────────────────
// the chain-timestamp recovery rule — chain-timestamp recovery from attestation key_id
// ─────────────────────────────────────────────────────────────────────────────

// RecoverTimestampFromKeyID recovers the chain timestamp from an attestation
// `key_id`. Mirrors `nanorix_verify::recover_timestamp_from_key_id`.
//
// AuditProofs issued before the chain-timestamp recovery rule restored the document-level `destroyed_at`
// field carry the chain timestamp in exactly one place: the attestation
// `key_id`, built by the chain specification as
// `nrx-verify-{terminated_at with ':' replaced by '-'}-{capsule_id[..8]}`.
// Only the TIME portion ever held colons — the ISO-8601 date carries its own
// dashes — so restoration splits at `T` and rewrites dashes on the right-hand
// side only. Fractional seconds and the zone suffix pass through untouched.
//
// Returns ok=false unless the reconstruction has the exact ISO-8601
// `YYYY-MM-DDTHH:MM:SS` shape; this function never guesses.
//
// Recovering from an attacker-mutable field is sound because `key_id` is
// covered by NEITHER signed message, so the recovered value is never trusted on
// its own — it is an INPUT to the chain walk, and the chain hashes it must
// reproduce ARE signature-bound. Exactly one timestamp string reproduces a
// signed chain, so a mutated `key_id` yields a mismatch and a rejection, never
// a false accept.
func RecoverTimestampFromKeyID(keyID string) (string, bool) {
	const prefix = "nrx-verify-"
	if len(keyID) < len(prefix) || keyID[:len(prefix)] != prefix {
		return "", false
	}
	rest := keyID[len(prefix):]

	// Strip the trailing `-{capsule_id[..8]}` fragment. Capsule-id fragments
	// are hex, so the LAST dash is the delimiter.
	lastDash := -1
	for i := len(rest) - 1; i >= 0; i-- {
		if rest[i] == '-' {
			lastDash = i
			break
		}
	}
	if lastDash < 0 {
		return "", false
	}
	encoded, fragment := rest[:lastDash], rest[lastDash+1:]
	if fragment == "" {
		return "", false
	}

	tIdx := -1
	for i := 0; i < len(encoded); i++ {
		if encoded[i] == 'T' {
			tIdx = i
			break
		}
	}
	if tIdx < 0 {
		return "", false
	}
	date, encodedTime := encoded[:tIdx], encoded[tIdx+1:]

	timePart := make([]byte, len(encodedTime))
	for i := 0; i < len(encodedTime); i++ {
		if encodedTime[i] == '-' {
			timePart[i] = ':'
		} else {
			timePart[i] = encodedTime[i]
		}
	}
	timeStr := string(timePart)

	if !isISO8601Shaped(date, timeStr) {
		return "", false
	}
	return date + "T" + timeStr, true
}

// isISO8601Shaped checks a `YYYY-MM-DD` date plus an `HH:MM:SS` time prefix.
// Anything after the seconds (fractional part, zone designator) is a free-form
// tail.
func isISO8601Shaped(date, timeStr string) bool {
	d, t := []byte(date), []byte(timeStr)
	if len(d) != 10 || len(t) < 8 {
		return false
	}
	if d[4] != '-' || d[7] != '-' {
		return false
	}
	for _, i := range [...]int{0, 1, 2, 3, 5, 6, 8, 9} {
		if d[i] < '0' || d[i] > '9' {
			return false
		}
	}
	if t[2] != ':' || t[5] != ':' {
		return false
	}
	for _, i := range [...]int{0, 1, 3, 4, 6, 7} {
		if t[i] < '0' || t[i] > '9' {
			return false
		}
	}
	return true
}

// resolveChainTimestamp resolves the timestamp every chain step hashes.
//
// Returns (timestamp, recovered) where `recovered` is non-nil only on the
// `key_id` recovery path — the caller records it in the verdict metadata so an
// auditor can always tell which route produced the result.
func resolveChainTimestamp(proof map[string]interface{}) (string, *string) {
	declared := stringOrEmpty(proof["destroyed_at"])
	if declared != "" {
		return declared, nil
	}
	keyID := lookupStringPath(proof, "attestation", "key_id")
	if keyID == nil {
		return declared, nil
	}
	ts, ok := RecoverTimestampFromKeyID(*keyID)
	if !ok {
		// No usable key_id — keep the pre-recovery behaviour exactly (an empty
		// timestamp, which fails the chain walk for any real proof).
		return declared, nil
	}
	return ts, strPtr(ts)
}

File: tools/nanorix-verify/src/canonical_recompute.rs
//! EO-07 sub-A — independent canonical-hash recompute + Ed25519 signature
//! verification for the single-proof path (`verify_auditproof`).
//!
//! ## Why this lives here (and not in a shared product crate)
//!
//! The offline verifier is deliberately INDEPENDENT of the product crates so
//! an auditor can compile and read it without pulling `services/api` /
//! `nanorix-rzl` and their transitive deps. `nanorix-verify-types` is a
//! pure-types crate (serde only) and cannot host the JCS + SHA-512 + Ed25519
//! runtime code. So — consistent with how this crate already mirrors
//! `nanorix_rzl::compute_step_hash` and the Wave-N primitives — we mirror the
//! server's canonical view here and lock byte-identity with the cross-impl
//! corpus (`fixtures/corpus/01_single_capsule_success` must verify; the
//! `04_failure_signature_invalid` and `08_failure_canonical_hash_drift`
//! categories must fail).
//!
//! ## What is mirrored
//!
//! `services/api/src/cdp_document.rs::FullCdp::canonical_view()` builds a
//! 16-field `CanonicalCdpView` (ADR-011 Part 3) and hashes it via
//! `nanorix_rzl::canonical::canonical_hash` = `hex(sha512(serde_jcs(view)))`.
//! Because the AuditProof JSON already contains every value in its exact
//! serialized shape, we rebuild only the *view* — mapping `FullCdp` wire field
//! names to the canonical-view keys and applying the two transforms the server
//! applies (`signing_key_version` String -> i64; the `attestation` subset) —
//! then re-canonicalize with the same RFC-8785 JCS. Under JCS the physical key
//! order is irrelevant (keys are sorted), so a byte-flip / key-reorder tamper
//! either changes a hashed value (signature fails) or is semantically identical
//! (correctly still verifies).
//!
//! ## Trust scope (sub-A vs sub-B)
//!
//! This verifies the signature against the public key EMBEDDED in the proof —
//! identical to what `services/api/src/routes/verify.rs` does today. That
//! detects tampering of a signed proof. It does NOT establish that the key
//! belongs to Nanorix; binding the key to a Nanorix-rooted trust anchor is
//! EO-07 sub-B (the trust-chain manifest in `trust_chain.rs`).

use crate::{strip_base64_prefix, strip_hash_prefix};
use base64::Engine;
use ed25519_dalek::{Signature, Verifier, VerifyingKey};
use nanorix_verify_types::SignatureFailureReason;
use serde_json::{Map, Value};
use sha2::{Digest, Sha512};

/// Marks a `signed_message` result as "this build cannot verify the declared
/// signing_mode" rather than an actual message to sign over. Never a valid hash.
const UNSUPPORTED_MODE_SENTINEL: &str = "\u{0}unsupported-signing-mode:";

/// Outcome of the signature stage (stages 5-8).
#[derive(Debug, PartialEq)]
pub enum SignatureCheck {
    /// Signature verified against the embedded key over the correct message.
    Verified,
    /// No signature present at all (unsigned partial). Nothing to check — the
    /// caller keeps the honest stage-4 "chain verified, signature NOT checked"
    /// verdict.
    Absent,
    /// The document declares a `signing_mode` this build cannot verify.
    ///
    /// Distinct from `Absent` on purpose. `signing_mode` is inside the canonical
    /// hash and is attacker-controllable, so if an unrecognised mode produced the
    /// same "signature not checked" outcome as a missing signature, flipping the
    /// mode would convert a rejection into a reassuring partial result — a
    /// downgrade oracle. "I have no signature to check" and "I cannot perform the
    /// verification this document requires" are different conditions and must
    /// produce different verdicts. Carries the offending mode.
    Unsupported(String),
    /// A signature was present but did not verify.
    Failed(SignatureFailureReason),
}

/// Rebuild the ADR-011 Part-3 canonical view from a proof's JSON and return its
/// RFC-8785 JCS SHA-512 hex digest — byte-identical to the server's
/// `FullCdp::canonical_hash()`. Lowercase 128-char hex (or empty on the
/// impossible JCS-serialize failure, which fails the signature check closed).
pub fn recompute_canonical_hash(proof: &Value) -> String {
    let str_field = |k: &str| proof.get(k).cloned().unwrap_or(Value::Null);
    let mut v: Map<String, Value> = Map::new();

    // Always-present scalars (FullCdp wire name -> canonical-view key).
    v.insert("version".into(), str_field("cdp_version"));
    v.insert("signing_mode".into(), str_field("signing_mode"));
    v.insert("jurisdiction".into(), str_field("jurisdiction"));
    v.insert("authority_id".into(), str_field("authority_id"));
    // signing_key_version: FullCdp stores a String; the canonical view emits an
    // integer (server parses, unparseable -> 0).
    let skv = proof
        .get("signing_key_version")
        .and_then(|x| x.as_str())
        .and_then(|x| x.parse::<i64>().ok())
        .unwrap_or(0);
    v.insert("signing_key_version".into(), Value::from(skv));
    v.insert("capsule_id".into(), str_field("capsule_id"));
    // org_id defaults to "" on the server (#[serde(default)] String).
    v.insert(
        "org_id".into(),
        proof
            .get("org_id")
            .cloned()
            .unwrap_or_else(|| Value::String(String::new())),
    );

    // skip_serializing_if = Option::is_none -> OMIT when absent/null.
    insert_if_present(&mut v, "parent_audit_proof_id", proof);
    insert_if_present(&mut v, "cdp_kind", proof);

    // Arrays carried verbatim (canonical-view key differs from wire name).
    v.insert(
        "activity_trail".into(),
        proof
            .get("activity")
            .cloned()
            .unwrap_or(Value::Array(vec![])),
    );
    v.insert(
        "destruction_chain".into(),
        proof.get("chain").cloned().unwrap_or(Value::Array(vec![])),
    );
    v.insert("destruction_state".into(), str_field("destruction_state"));

    // No skip attribute -> serialized as null when absent (NOT omitted).
    v.insert(
        "destruction_failure_step".into(),
        str_field("destruction_failure_step"),
    );

    insert_if_present(&mut v, "parent_proofs_merkle_root", proof);
    insert_if_present(&mut v, "record_receipts_merkle_root", proof);

    // No skip attribute -> null when absent.
    v.insert(
        "runtime_attestation".into(),
        str_field("runtime_attestation"),
    );

    // attestation subset: { timestamp_attestation (null|obj),
    // attestation_chain_fingerprint (null|str — empty string canonicalizes to
    // null, matching the server's `if fingerprint.is_empty() { None }`). }
    let mut att: Map<String, Value> = Map::new();
    att.insert(
        "timestamp_attestation".into(),
        str_field("timestamp_attestation"),
    );
    let fingerprint = proof
        .get("attestation_chain_fingerprint")
        .and_then(|x| x.as_str())
        .filter(|s| !s.is_empty());
    att.insert(
        "attestation_chain_fingerprint".into(),
        fingerprint.map(Value::from).unwrap_or(Value::Null),
    );
    v.insert("attestation".into(), Value::Object(att));

    v.insert("hash_algorithm".into(), str_field("hash_algorithm"));
    v.insert(
        "signature_algorithm".into(),
        str_field("signature_algorithm"),
    );

    match serde_jcs::to_vec(&Value::Object(v)) {
        Ok(bytes) => {
            let digest = Sha512::digest(&bytes);
            digest.iter().map(|b| format!("{:02x}", b)).collect()
        }
        // Object always serializes; on the impossible failure return empty so
        // the signature comparison fails closed rather than panicking.
        Err(_) => String::new(),
    }
}

fn insert_if_present(view: &mut Map<String, Value>, key: &str, proof: &Value) {
    if let Some(x) = proof.get(key) {
        if !x.is_null() {
            view.insert(key.to_string(), x.clone());
        }
    }
}

/// Select the signed message for a proof by version/mode — mirrors
/// `services/api/src/routes/verify.rs`:
/// - `1.0` -> `final_hash` (ASCII hex, prefix-stripped)
/// - `2.0` -> `document_hash`
/// - `2.1` + `nanorix_only` -> recomputed `canonical_hash`
/// - `2.1` + `dual_signature` / `tee_attested` -> `None` (not verifiable here)
fn signed_message(proof: &Value, cdp_version: &str) -> Option<String> {
    let signing_mode = proof
        .get("signing_mode")
        .and_then(|v| v.as_str())
        .unwrap_or("nanorix_only");
    let msg = match cdp_version {
        "1.0" => strip_hash_prefix(
            proof
                .get("final_hash")
                .and_then(|v| v.as_str())
                .unwrap_or(""),
        )
        .to_string(),
        "2.0" => strip_hash_prefix(
            proof
                .get("document_hash")
                .and_then(|v| v.as_str())
                .unwrap_or(""),
        )
        .to_string(),
        "2.1" => match signing_mode {
            "nanorix_only" => recompute_canonical_hash(proof),
            // Any other declared mode is one this build cannot verify. It is NOT
            // "no signature" — see SignatureCheck::Unsupported. Signalled with a
            // sentinel the callers translate, so this fn keeps its Option shape.
            other => return Some(format!("{UNSUPPORTED_MODE_SENTINEL}{other}")),
        },
        _ => return None,
    };
    Some(msg)
}

/// Decode base64 Ed25519 signature + public key and verify `message` under
/// them. Shared by the embedded-key (sub-A) and manifest-key (sub-B) paths,
/// and by the ADR-050 BoundaryAttestation sibling pipeline (`boundary.rs`).
pub(crate) fn verify_message_with_key(
    message: &str,
    sig_b64: &str,
    pub_b64: &str,
) -> SignatureCheck {
    let sig_bytes =
        match base64::engine::general_purpose::STANDARD.decode(strip_base64_prefix(sig_b64)) {
            Ok(b) if b.len() == 64 => b,
            _ => return SignatureCheck::Failed(SignatureFailureReason::Malformed),
        };
    let pub_bytes =
        match base64::engine::general_purpose::STANDARD.decode(strip_base64_prefix(pub_b64)) {
            Ok(b) if b.len() == 32 => b,
            _ => return SignatureCheck::Failed(SignatureFailureReason::PublicKeyMalformed),
        };
    let sig_array: [u8; 64] = match sig_bytes.as_slice().try_into() {
        Ok(a) => a,
        Err(_) => return SignatureCheck::Failed(SignatureFailureReason::Malformed),
    };
    let pub_array: [u8; 32] = match pub_bytes.as_slice().try_into() {
        Ok(a) => a,
        Err(_) => return SignatureCheck::Failed(SignatureFailureReason::PublicKeyMalformed),
    };
    let signature = Signature::from_bytes(&sig_array);
    let verifying_key = match VerifyingKey::from_bytes(&pub_array) {
        Ok(k) => k,
        Err(_) => return SignatureCheck::Failed(SignatureFailureReason::PublicKeyMalformed),
    };
    match verifying_key.verify(message.as_bytes(), &signature) {
        Ok(()) => SignatureCheck::Verified,
        Err(_) => SignatureCheck::Failed(SignatureFailureReason::DoesNotVerify),
    }
}

/// The proof's embedded attestation signature (base64), if present + non-empty.
fn embedded_signature(proof: &Value) -> Option<&str> {
    proof
        .pointer("/attestation/signature")
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
}

/// sub-A — verify the proof's signature against the public key EMBEDDED in the
/// proof's attestation. Proves integrity (not tampered since signing), NOT
/// authenticity. Returns `Absent` when no signature/key is present or the
/// version/mode is not signature-verifiable here.
pub fn verify_signature(proof: &Value, cdp_version: &str) -> SignatureCheck {
    let pub_b64 = proof
        .pointer("/attestation/public_key")
        .or_else(|| proof.pointer("/attestation/verification_key"))
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty());
    let (sig_b64, pub_b64) = match (embedded_signature(proof), pub_b64) {
        (Some(s), Some(p)) => (s, p),
        _ => return SignatureCheck::Absent,
    };
    let message = match signed_message(proof, cdp_version) {
        Some(m) => match m.strip_prefix(UNSUPPORTED_MODE_SENTINEL) {
            Some(mode) => return SignatureCheck::Unsupported(mode.to_string()),
            None => m,
        },
        None => return SignatureCheck::Absent,
    };
    verify_message_with_key(&message, sig_b64, pub_b64)
}

/// sub-B — verify the proof's signature against a trust-chain-RESOLVED public
/// key (from the signed manifest), NOT the embedded one. This is what
/// establishes authenticity: a forged proof carrying its own embedded key
/// passes sub-A but fails here, because its key is not the manifest key.
/// Returns `Absent` when the proof has no signature or is not verifiable here.
pub fn verify_signature_against(proof: &Value, cdp_version: &str, pub_b64: &str) -> SignatureCheck {
    let sig_b64 = match embedded_signature(proof) {
        Some(s) => s,
        None => return SignatureCheck::Absent,
    };
    let message = match signed_message(proof, cdp_version) {
        Some(m) => match m.strip_prefix(UNSUPPORTED_MODE_SENTINEL) {
            Some(mode) => return SignatureCheck::Unsupported(mode.to_string()),
            None => m,
        },
        None => return SignatureCheck::Absent,
    };
    verify_message_with_key(&message, sig_b64, pub_b64)
}

#[cfg(test)]
mod tests {
    use super::*;
    use base64::Engine;
    use ed25519_dalek::{Signer, SigningKey};
    use serde_json::json;

    /// The wire form of `services/api/src/cdp_document.rs::test_cdp()` — the
    /// fixed input behind the server's `golden_canonical_hash` test. Only the
    /// fields the canonical view actually reads are present; excluded fields
    /// (residency, final_hash, destruction_trigger, the main attestation, the
    /// reserved slots) are omitted because the view never reads them.
    fn golden_input() -> serde_json::Value {
        json!({
            "cdp_version": "2.1",
            "capsule_id": "cap_test_0001",
            "signing_key_version": "1",
            "activity": [],
            "chain": [],
            "signing_mode": "nanorix_only",
            "jurisdiction": "us",
            "authority_id": "us-kms-nanorix-v1",
            "org_id": "00000000-0000-0000-0000-000000000001",
            "destruction_state": "complete",
            "hash_algorithm": "SHA-512",
            "signature_algorithm": "Ed25519"
        })
    }

    /// BYTE-EXACTNESS LOCK against the signer. The offline recompute must equal
    /// the server's golden `402f533e…` (`cdp_document.rs::golden_canonical_hash`)
    /// for the same input. If the server's canonical view ever changes this
    /// digest changes too and BOTH goldens must be updated in lockstep — this
    /// is the cross-crate guard that keeps the independent offline verifier from
    /// silently drifting away from how proofs are actually signed.
    #[test]
    fn canonical_recompute_matches_server_golden() {
        assert_eq!(
            recompute_canonical_hash(&golden_input()),
            "402f533e81d78a05f386bee62a919436b8eacdea2c49397d54547bbd19dabce47bc95143635ee6d487e35eee037fec31ebd1f8f88b2f3f36ae2908324c74aabe",
            "offline canonical recompute drifted from the server golden"
        );
    }

    /// Sign-then-verify roundtrip over the recomputed canonical hash (the
    /// production v2.1 `nanorix_only` message). Proves (1) a correctly signed
    /// proof verifies, (2) flipping a canonical-bound field (jurisdiction — the
    /// exact corpus `08` drift case) is caught even though it does NOT touch the
    /// 8-step chain / `final_hash` (the property a final_hash-only verifier
    /// would MISS), and (3) an unsigned proof reports `Absent`, never a false
    /// pass.
    #[test]
    fn roundtrip_verify_and_catch_canonical_drift() {
        let key = SigningKey::from_bytes(&[7u8; 32]);
        let canonical = recompute_canonical_hash(&golden_input());
        let sig = key.sign(canonical.as_bytes());

        let mut signed = golden_input();
        signed["attestation"] = json!({
            "algorithm": "Ed25519",
            "public_key": base64::engine::general_purpose::STANDARD
                .encode(key.verifying_key().to_bytes()),
            "signature": base64::engine::general_purpose::STANDARD.encode(sig.to_bytes()),
        });
        assert_eq!(verify_signature(&signed, "2.1"), SignatureCheck::Verified);

        let mut drifted = signed.clone();
        drifted["jurisdiction"] = json!("eu");
        assert!(matches!(
            verify_signature(&drifted, "2.1"),
            SignatureCheck::Failed(_)
        ));

        assert_eq!(
            verify_signature(&golden_input(), "2.1"),
            SignatureCheck::Absent
        );
    }
}
